test/connect.cpp

```diff
@@ -230,15 +230,15 @@ BOOST_AUTO_TEST_CASE( keep_alive_and_send_control_packet ) {
                     BOOST_TEST(connack_return_code == MQTT_NS::connect_return_code::accepted);
                     tim.expires_from_now(boost::posix_time::seconds(2));
                     tim.async_wait(
-                        [&chk, &c, &tim](boost::system::error_code const& ec) {
+                        [&chk, &c, &tim](std::error_code ec) {
                             MQTT_CHK("2sec");
                             BOOST_CHECK(!ec);
                             c->publish("topic1", "timer_reset", MQTT_NS::qos::at_most_once);
                             tim.expires_from_now(boost::posix_time::seconds(4));
                             tim.async_wait(
-                                [&chk](boost::system::error_code const& ec) {
+                                [&chk](std::error_code ec) {
```

Comment:
Why do you change the error code? 
`mqtt_cpp` depends on boost. It is natural that mqtt_cpp use `boost::system::error_code`.

Is boost asio support `std::error_code` ?
I couldn't find it in the document:
https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/reference/basic_deadline_timer/async_wait.html

Reply:
boost::system::error_code implicitly converts to std::error_code.

https://www.boost.org/doc/libs/1_66_0/boost/system/error_code.hpp

```
#ifndef BOOST_NO_CXX11_HDR_SYSTEM_ERROR

      operator std::error_code () const BOOST_SYSTEM_NOEXCEPT
      {
        return std::error_code( value(), category() );
      }

#endif
```


std::error_code is less typing.

Reply:
> boost::system::error_code implicitly converts to std::error_code.

Good to know, thanks.

> std::error_code is less typing.

True.
Is this change doesn't break existing code?
Can `boost::system::error_code` create from `std::error_code` ?

Reply:
How about exception? Boost.Asio throws exception. And from the exception, we can get `boost::system::error_code`. I think that `boost::system::error_code` is better for consistency.

If the reason is less typing, simply we can use `using some_alias = boost::system::error_code`. It doesn't break user's code.

Reply:
I do not know if std::error_code implicltly converts to boost::system::error_code.

We can wait for next release to break API

Reply:
> I do not know if std::error_code implicltly converts to boost::system::error_code.
> 
> We can wait for next release to break API

I don't see enough reason to replace `boost::system::error_code` with `std::error_code`, so far.

I mentioned one problem that is consistency with exception, and one alternative solution that uses type alias.

Reply:
Where does boost asio throw the exception?

Reply:
For example, mqtt_cpp calls https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/reference/connect/overload3.html

Reply:
I decided that mqtt_cpp keep `boost::system::error_code`.
Here are the reasons.

1. `boost::system_error_code` can convert to `std::error_code` but not vice versa. See https://wandbox.org/permlink/WGdr2yqqtldSZZUg. If mqtt_cpp user wants to use `std::error_code`, they can write wrapper easily. However, if `boost::system_error_code` is replaced with `std::error_code`, existing user codes would be broken and writing wrapper is not so simple.
2. mqtt_cpp is Boost.Asio based library. So users's expectation is respect Boost.Asio's mannar. `std::error_code` is something surprising.
3. mqtt_cpp uses Boost.Asio synchronous APIs that throw exception. It is very simple and natural way. We can catch the exception and convert to `std` one and rethrow it. but it is not natural. We can use error_code version of Asio APIs and create `std` exception from the error code. It is also unnatural and not simple.

However, C++ standard network library that is based on Boost.Asio is in under proposal. If C++ standard would have it in the future, then we can replace `std` one. It is very natural.

Reply:
Sure, that's fine.

Reply:
@redboltz How about

```cpp
namespace MQTT_NS
{
    using error_code = boost::system::error_code;
} // namespace MQTT_NS
```

and then using `MQTT_NS::error_code` instead of `std::error_code` in this PR ?